### PR TITLE
Group Voting, fixes to testing

### PR DESCRIPTION
**Pull request by patimen**

### CODE REVIEW

File: decentralized-api/event_listener.go [hunk 1]
```diff
@@ -67,16 +67,58 @@ func StartEventListener(nodeBroker *broker.Broker, transactionRecorder cosmoscli
...
+}
+
+func waitForEventHeight(event chainevents.JSONRPCResponse) bool {
+	heightString := event.Result.Events["tx.height"][0]
+	expectedHeight, err := strconv.ParseInt(heightString, 10, 64)
+	if err != nil {
+		slog.Error("Failed to parse height", "error", err)
+		return true
+	}
+	for poc.CurrentHeight < expectedHeight {
```
Comment (patimen) [inline decentralized-api/event_listener.go:107]: CurrentHeight is just a primitive, should probably be Atomic at some point.

File: decentralized-api/validation.go [hunk 2]
```diff
@@ -17,28 +18,74 @@ import (
 	"math"
 	"math/rand"
 	"net/http"
+	"strconv"
 	"strings"
 )
 
+var revalidated = make(map[string]int)
```
Comment (patimen) [inline decentralized-api/validation.go]: This likely is not longer needed, as we only catch events with needs_revalidation.

File: inference-chain/x/inference/keeper/msg_server_re_validation.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/productscience/inference/x/inference/types"
+)
+
+func (k msgServer) ReValidation(goCtx context.Context, msg *types.MsgReValidation) (*types.MsgReValidationResponse, error) {
```
Comment (patimen) [inline inference-chain/x/inference/keeper/msg_server_re_validation.go]: This was added by accident, should be removed.

File: inference-chain/x/inference/keeper/msg_server_validation.go [hunk 4]
```diff
@@ -32,42 +38,88 @@ func (k msgServer) Validation(goCtx context.Context, msg *types.MsgValidation) (
...
+	voteMsg.Metadata = "Revalidate inference " + inference.InferenceId
+	err = k.vote2(ctx, voteMsg)
+	if err != nil {
+		return nil, err
+	}
 	return &types.MsgValidationResponse{}, nil
 }
+func (k msgServer) vote2(ctx sdk.Context, vote *group.MsgVote) error {
+	k.LogInfo("Validation: Voting", "vote", vote)
+	_, err := k.group.Vote(ctx, vote)
```
Comment (patimen) [inline inference-chain/x/inference/keeper/msg_server_validation.go:111]: THIS is where the actual messsage to validate/revalidate will execute.

File: inference-chain/x/inference/keeper/msg_server_invalidate_inference.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+	k.SetInference(ctx, inference)
+	k.SetParticipant(ctx, executor)
+	return &types.MsgInvalidateInferenceResponse{}, nil
+}
+
+func (k msgServer) markInferenceAsInvalid(executor *types.Participant, inference *types.Inference, ctx sdk.Context) error {
+	inference.Status = types.InferenceStatus_INVALIDATED
+	executor.InvalidatedInferences++
+	executor.CoinBalance -= inference.ActualCost
```
Comment (DimaOrekhovPS) [inline inference-chain/x/inference/keeper/msg_server_invalidate_inference.go:49]: What if the invalidation happens after the reckoning period? Or we don't release the rewards for inferences that haven't been validated yet?

File: launch_chain.sh [hunk 6]
```diff
@@ -43,6 +43,11 @@ if [ -z "$SEED_IP" ]; then
   exit 1
 fi
 
+if [ -z "$EXTERNAL_SEED_IP" ]; then
```
Comment (DimaOrekhovPS) [inline launch_chain.sh:46]: What's the difference?

**Review by DimaOrekhovPS — APPROVED** [review]

**Pull request merged**